state/src/applicator/task_queue.rs

```diff
@@ -537,24 +544,39 @@ impl StateApplicator {
             task.state = QueuedTaskState::Failed;
         }
 
-        // Remove the task's node assignment and add it to history
+        // Add the task to history and remove its node assignment
+        self.maybe_append_historical_task(*key, task.clone(), tx)?;
         if let Some(executor) = tx.get_task_assignment(&task.id)? {
             tx.remove_assigned_task(&executor, &task.id)?;
         }
-        Self::maybe_append_historical_task(*key, task.clone(), tx)?;
         Ok(Some(task))
     }
 
     /// Append a task to the task history if it should be stored
     fn maybe_append_historical_task(
+        &self,
         key: TaskQueueKey,
         task: QueuedTask,
         tx: &StateTxn<'_, RW>,
     ) -> Result<()> {
         if let Some(t) = HistoricalTask::from_queued_task(key, task)
             && tx.get_historical_state_enabled()?
         {
-            tx.append_task_to_history(&key, t)?;
+            tx.append_task_to_history(&key, t.clone())?;
+
+            // Emit a task completion event to the event manager
+            // _only if the local peer is the executor_,
+            // to avoid duplicate events across the cluster
+            let my_peer_id = tx.get_peer_id()?;
+            let maybe_executor = tx.get_task_assignment(&t.id)?;
+            let is_executor = maybe_executor.is_some_and(|e| e == my_peer_id);
+
+            if is_executor {
+                let event = RelayerEvent::TaskCompletion(TaskCompletionEvent::new(key, t));
+                if let Err(e) = self.config.event_queue.send(event) {
+                    error!("error sending task completion event: {e}");
+                }
+            }
```

Comment:
shouldn't this happen outside the `if` block here? we will have `tx.get_historical_state_enabled() -> false` for our nodes internally.

Reply:
great catch